### PR TITLE
Example/mp4sample supports fragmented mp4

### PR DESCRIPTION
**Pull request by nintha**
1. Example/mp4sample supports fragmented mp4.
2. Add some fields into trun box :
    - first_sample_flags
    - sample_durations 
    - sample_flags
    - sample_cts
3. And set correct flag in trun box.

### CODE REVIEW

File: src/track.rs [hunk 1]
```diff
@@ -363,24 +363,28 @@ impl Mp4Track {
...
-                } else {
-                    return Err(Error::EntryInTrunNotFound(
-                        self.track_id(),
-                        BoxType::TrunBox,
-                        sample_id,
-                    ));
+            let mut offset = 0;
+            for traf_idx in 0..self.trafs.len() {
+                if let Some(trun) = &self.trafs[traf_idx].trun {
+                    let sample_count = trun.sample_count;
```
Comment (alfg) [inline src/track.rs:369]: Possible to utilize `self.sample_count()`?

File: src/track.rs [hunk 2]
```diff
@@ -413,9 +417,20 @@ impl Mp4Track {
 
     fn sample_offset(&self, sample_id: u32) -> Result<u64> {
         if self.trafs.len() > 0 {
-            let sample_sizes_count = self.sample_count() / self.trafs.len() as u32;
-            let traf_idx = (sample_id - 1) / sample_sizes_count;
-            Ok(self.trafs[(sample_id - (sample_sizes_count * traf_idx)) as usize].tfhd.base_data_offset as u64)
+            let mut offset = 0;
+            for traf_idx in 0..self.trafs.len() {
+                if let Some(trun) = &self.trafs[traf_idx].trun {
+                    let sample_count = trun.sample_count;
```
Comment (alfg) [inline src/track.rs:423]: Also here, if possible. Unless it can't be used here.

**Review by alfg — CHANGES_REQUESTED** [review]
Hey @nintha, 

Thanks again for the fix. I tried running the `mp4sample` on one of my test fragmented example and got the following issue at the end:

```
...
[14997] start_time=14996000 duration=1000 rendering_offset=0 size=114 is_sync=false
[14998] start_time=14997000 duration=1000 rendering_offset=0 size=124 is_sync=false
[14999] start_time=14998000 duration=1000 rendering_offset=0 size=105 is_sync=false
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Custom { kind: UnexpectedEof, error: "failed to fill whole buffer" })', examples\mp4sample.rs:37:37
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: process didn't exit successfully: `target\debug\examples\mp4sample.exe fragmented.mp4` (exit code: 101)
```

Can you check into this? I've attached the reference file. Thanks.

https://user-images.githubusercontent.com/702541/106077835-2296d200-60c7-11eb-8a0f-1ea337c0e1c6.mp4

**Comment by nintha** [comment]
Hi @alfg
Thanks for your comments, and I try to fix it :
1. Field data_offset in TrunBox is optional. The `fragmented.mp4` does not have data_offset in TrunBox.  
2. Extract duplicated codes as a  new function.

Now `mp4sample` works well with `fragmented.mp4`.

**Comment by alfg** [comment]
@nintha Thanks for fixing!

**Pull request merged**